pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,885 @@
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if tt.wantSyncKeeperErr {
+				if err := ehc.sendToKeepChannel(*tt.message); (err != nil) != tt.wantErr {
+					t.Errorf("EdgeHubController.sendToKeepChannel() error = %v, wantErr %v", err, tt.wantErr)
+				}
+			} else if tt.wantErr {
+				ehc.addKeepChannel("test_id")
+				if err := ehc.sendToKeepChannel(*tt.message); (err != nil) != tt.wantErr {
+					t.Errorf("EdgeHubController.sendToKeepChannel() error = %v, wantErr %v", err, tt.wantErr)
+				}
+			} else {
+				keepChannel := ehc.addKeepChannel("test_id")
+				go ehc.sendToKeepChannel(*tt.message)
+				got := <-keepChannel
+				if got != *tt.message {
+					t.Errorf("SendToKeep() error want: %v got: %v", *tt.message, got)
+				}
+			}
+
+		})
+	}
+}
+
+//TestDispatch() tests whether the messages are properly dispatched to their respective modules
+func TestDispatch(t *testing.T) {
+	tests := []struct {
+		name       string
+		fields     EdgeHubController
+		message    *model.Message
+		wantErr    bool
+		isResponse bool
+	}{
+		{"TestDispatch: Positive Test Case", EdgeHubController{
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			syncKeeper: make(map[string]chan model.Message),
+		}, model.NewMessage("").BuildRouter(ModuleNameEdgeHub, core.TwinGroup, "", ""),
+			false, false},
+
+		{"TestDispatch: Error Case", EdgeHubController{
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			syncKeeper: make(map[string]chan model.Message),
+		}, model.NewMessage("test").BuildRouter(ModuleNameEdgeHub, core.EdgedGroup, "", ""),
+			true, true},
+
+		{"TestDispatch: Response Case", EdgeHubController{
+			context:    context.GetContext(context.MsgCtxTypeChannel),
+			syncKeeper: make(map[string]chan model.Message),
+		}, model.NewMessage("test").BuildRouter(ModuleNameEdgeHub, core.TwinGroup, "", ""),
+			false, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ehc := &EdgeHubController{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if tt.isResponse && tt.wantErr != true {
+				keepChannel := ehc.addKeepChannel("test")
+				go ehc.dispatch(*tt.message)
+				got := <-keepChannel
+				if got != *tt.message {
+					t.Errorf("SendToKeep() error want: %v got: %v", *tt.message, got)
+				}
+			} else {
+				if err := ehc.dispatch(*tt.message); (err != nil) != tt.wantErr {
+					t.Errorf("EdgeHubController.dispatch() error = %v, wantErr %v", err, tt.wantErr)
+				}
+			}
+		})
+	}
+}
+
+//TestRouteToEdge() is used to test whether the message received from websocket is dispatched to the required modules
+func TestRouteToEdge(t *testing.T) {
+	initMocks(t)
+	tests := []struct {
+		name       string
+		fields     EdgeHubController
+		receiveErr bool
```

Comment:
receiveErr -> receiveBool ??

Reply:
Here receiveErr variable is used to signify whether it is an error in receiving the message from the cloud. It basically is used to tell the user whether the test case is expecting an error or not. If receiveErr is true, it means that an error is expected in the test case, if it is false then it signifies that no error is expected in the reception of the message. I feel that receiveBool might not convey the message properly and might not be easy to understand.